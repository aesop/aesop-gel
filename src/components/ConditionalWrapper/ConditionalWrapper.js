import React from 'react';
import PropTypes from 'prop-types';

const ConditionalWrapper = ({
  alternateWrapper,
  children,
  condition,
  wrapper,
}) => {
  if (condition) return wrapper(children);

  return alternateWrapper ? alternateWrapper(children) : <>{children}</>;
};

ConditionalWrapper.propTypes = {
  /**
   * If the condition false, this functional component will be used instead of simply returning the childen unwrapped.
   */
  alternateWrapper: PropTypes.any,
  children: PropTypes.node.isRequired,
  /**
   * If this boolean prop resolves to true the children prop will be wrapped in the wrapper prop.
   */
  condition: PropTypes.bool.isRequired,
  /**
   * A functional component that recieves the children prop as an argument, `children => <div>{children}</div>`
   */
  wrapper: PropTypes.any.isRequired,
};

ConditionalWrapper.defaultProps = {
  alternateWrapper: undefined,
  children: undefined,
  condition: undefined,
  wrapper: undefined,
};

export default ConditionalWrapper;
