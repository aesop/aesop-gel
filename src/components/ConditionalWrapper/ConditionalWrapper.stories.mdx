import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import ConditionalWrapper from './ConditionalWrapper';
import ConditionalWrapperFixture from './ConditionalWrapper.fixture';

<Meta title="Components/ConditionalWrapper" component={ConditionalWrapper} />

# ConditionalWrapper

<Preview>
  <Story name="Base component">
    <ConditionalWrapper
      condition={true}
      wrapper={children => <div>{children}</div>}
    >
      <span>Conditionally wrapped content</span>
    </ConditionalWrapper>
  </Story>
</Preview>

## API

<Props of={ConditionalWrapper} />
