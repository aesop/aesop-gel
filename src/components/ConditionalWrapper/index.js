/* istanbul ignore file */
import ConditionalWrapper from './ConditionalWrapper';

export default ConditionalWrapper;
