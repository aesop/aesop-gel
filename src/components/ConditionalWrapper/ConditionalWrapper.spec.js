import React from 'react';
import renderer from 'react-test-renderer';
import ConditionalWrapper from './ConditionalWrapper';

describe('<ConditionalWrapper />', () => {
  it('should be defined', () => {
    expect(ConditionalWrapper).toBeDefined();
  });

  it('renders base component correctly', () => {
    const tree = renderer
      .create(
        <ConditionalWrapper
          condition={true}
          wrapper={children => <div className="wrapper">{children}</div>}
        >
          <span className="children">children</span>
        </ConditionalWrapper>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders children only is condition is false', () => {
    const tree = renderer
      .create(
        <ConditionalWrapper
          condition={false}
          wrapper={children => <div className="wrapper">{children}</div>}
        >
          <span className="children">children</span>
        </ConditionalWrapper>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders alternateWrapper correctly', () => {
    const tree = renderer
      .create(
        <ConditionalWrapper
          alternateWrapper={children => (
            <div className="alternateWrapper">{children}</div>
          )}
          condition={false}
          wrapper={children => <div className="wrapper">{children}</div>}
        >
          <span className="children">children</span>
        </ConditionalWrapper>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
