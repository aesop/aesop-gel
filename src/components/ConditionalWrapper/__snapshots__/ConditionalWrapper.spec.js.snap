// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConditionalWrapper /> renders alternateWrapper correctly 1`] = `
<div
  className="alternateWrapper"
>
  <span
    className="children"
  >
    children
  </span>
</div>
`;

exports[`<ConditionalWrapper /> renders base component correctly 1`] = `
<div
  className="wrapper"
>
  <span
    className="children"
  >
    children
  </span>
</div>
`;

exports[`<ConditionalWrapper /> renders children only is condition is false 1`] = `
<span
  className="children"
>
  children
</span>
`;
